### ISSUE TITLE
Kotlin inline function shows branch missing error

### ISSUE THREAD

**Issue by salmanjawedfolio** [jacoco/jacoco#1168]
Hello Jacoco team

I wanted to submit an issue

JaCoCo version : Created with JaCoCo 0.8.5.201910111838
OS : Mac OS catalina 10.15.7
Tool Integration: Gradle on Android Studio

if (objResponse.result is HNPGetListTemplatesResult) {
        val templateList = (objResponse.result as HNPGetListTemplatesResult).templates (Missed here)
}

The error is shown as 1 of 2 branches missed.
Same for the following code:
fun getInstance(hldbTemplate: HLDBTemplate, templatesEventAction: HLGetTemplatesEventAction) =
                instance ?: synchronized(this) {
                    instance ?: HLTemplatesRepository(hldbTemplate, templatesEventAction).also { instance = it }} (Missed here)

This should not be reporting branches missed according to my understanding and implementation. Can you please check and get back to me. Thank you

**Comment by salmanjawedfolio** [comment]
@Godin hello. can you please provide details?

**Comment by Godin** [comment]
@salmanjawedfolio you use `?:` so called "elvis operator" ( https://kotlinlang.org/docs/null-safety.html#elvis-operator )

> If the expression to the left of ?: is not null, the elvis operator returns it, otherwise it returns the expression to the right

So it introduces branch - your code is equivalent to the following

```
        fun getInstance() =
                if (instance == null) {
                    synchronized(this) {
                        if (instance == null) {
                            HLTemplatesRepository(hldbTemplate, templatesEventAction).also { instance = it }
                        } else {
                            instance
                        }
                    }
                } else {
                    instance
                }
```

where you can clearly see your uncovered branch

![example](https://user-images.githubusercontent.com/138671/112299014-acae8180-8c97-11eb-8c8e-f22abc589a74.png)

**Comment by salmanjawedfolio** [comment]
@Godin what about this case
if (objResponse.result is HNPGetListTemplatesResult) {
        val templateList = (objResponse.result as HNPGetListTemplatesResult).templates (Missed here)
}

**Comment by Godin** [comment]
> what about this case
> 
> ```
> if (objResponse.result is HNPGetListTemplatesResult) {
> val templateList = (objResponse.result as HNPGetListTemplatesResult).templates (Missed here)
> }
> ```

probably you are using Kotlin version 1.4 or never - #1134 ,
filter for unsafe-cast operators was improved for Kotlin 1.4 and later in not yet released JaCoCo version 0.8.7 - #1143

**Comment by salmanjawedfolio** [comment]
@Godin thanks a lot for the explanation. So based on my understanding the above issue will be fixed with later versions of JaCoCo right?
